### PR TITLE
Add ZZ[i] and QQ[i]

### PR DESCRIPTION
**Pull request by tthsqe12**
context: https://github.com/Nemocas/Nemo.jl/issues/1110, https://github.com/Nemocas/Nemo.jl/issues/1072
I would be nice to have a ZZ[i] and QQ[i] that embed directly (i.e. without user intervention) into the complex numbers.
If you are asking yourself if we really need this, consider the following:
1. `BigInt` is to `fmpz` as `Complex{BigInt}` is to `fmpzi`.
2. `Rational{BigInt}` is to `fmpq` as `Complex{Rational{BigInt}}` is to `fmpqi`.
3. `fmpzi` has a Euclidean structure that we should be able to use
4. Unlike `nf_elem`, the `fmpzi` and `fmpqi` embed directly into the complex numbers so that conversions and roundings between `acb` are possible.

If the ZZi and QQi are completely unacceptable, please let me know and I will stop. Otherwise, I will finish the rest of the interface. Also, @thofma, I don't care any more what `numerator(::fmpqi)` and `denominator(::fmpqi)` do. They can do whatever you want. And, I don't even care if this is changed:
```julia
base_ring(a::FlintQQiField) = ZZi
```

### CODE REVIEW

**Comment by tthsqe12** [comment]
@albinahlback, do you have ideas for the canonicalization of the gcdx coefficients in ZZ[i]?

**Comment by albinahlback** [comment]
I have a feeling that a generalized Euclidean algorithm should return the canonical coefficients. What I mean is that via [these definitions](https://en.wikipedia.org/wiki/Gaussian_integer#Euclidean_division), one should be able to generalize [this proof for canonical Bézout coefficients](https://math.stackexchange.com/questions/670405/does-the-extended-euclidean-algorithm-always-return-the-smallest-coefficients-of).

**Comment by tthsqe12** [comment]
I was thinking less of having the canonicity defined by an algorithm and more of having it defined by a set of easy-to-check conditions, like the flint fmpz docs. I already have the gcd satsifying pi/4 < gcd(,) <= pi/4. so that the gcd is as close the the real axis as possible.

**Comment by albinahlback** [comment]
Ah, yeah. Let me think of a definition.

**Comment by albinahlback** [comment]
I think the easiest way to deal with this is to assume that the numbers are coprime (in sense of Gaussian integers).

Let `a` and `b` be coprime Gaussian integers where `|a|^2, |b|^2 > 1`. Then I am guessing that we should be able to find `x` and `y` satisfying `a x + b y = 1` such that `|x|^2 \le |b|^2 / 2` and `|y|^2 \le |a|^2`.

And if `|a|^2 = 1`, then `x = 1 / a` and `y = 0` (and vice versa).

This should be canonical Bézout coefficients. However, if this holds along with that they are unique, I do not know.

**Comment by fingolfin** [comment]
Thank you for working on this. I am fine with doing this in principle. As long as there is a good way to convert between these new elements, and existing number field elements?

Also, supporting all the features we have for the isomorphic existing field types. Like, can we also factorize in polynomial rings over QQ[i]?

**Comment by tthsqe12** [comment]
Yes, I am discussing with thofma how to make it work as a number field/ring.

File: src/gaussiannumbers/QQi.jl [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+end
+
+function deepcopy_internal(a::FlintQQiField, d::IdDict)
+   return a
+end
+
+function deepcopy(a::FlintQQiField)
+   return a
+end
+
```
Comment (thofma) [inline src/gaussiannumbers/QQi.jl:178]:
```suggestion
```

File: src/gaussiannumbers/ZZi.jl [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+function deepcopy_internal(a::FlintZZiRing, d::IdDict)
+   return a
+end
+
+function deepcopy(a::FlintZZiRing)
+   return a
+end
+
+function real(a::fmpzi)
+   return deepcopy(a.x)
```
Comment (thofma) [inline src/gaussiannumbers/ZZi.jl:151]: Why the deepcopy?

File: src/gaussiannumbers/ZZi.jl [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+end
+
+function deepcopy_internal(a::FlintZZiRing, d::IdDict)
+   return a
+end
+
+function deepcopy(a::FlintZZiRing)
+   return a
+end
+
```
Comment (thofma) [inline src/gaussiannumbers/ZZi.jl:149]:
```suggestion
```

File: src/gaussiannumbers/ZZi.jl [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+function real(a::fmpzi)
+   return deepcopy(a.x)
+end
+
+function imag(a::fmpzi)
+   return deepcopy(a.y)
```
Comment (thofma) [inline src/gaussiannumbers/ZZi.jl:155]: Why the deepcopy?

File: src/gaussiannumbers/ZZi.jl [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+end
+
+function conj(a::fmpzi)
+   return fmpzi(deepcopy(a.x), -a.y)
```
Comment (thofma) [inline src/gaussiannumbers/ZZi.jl:159]: As above

**Review by thofma — COMMENTED** [review]
There are a bunch of `deepcopy`, like in `conj`, which I am not sure about.

According to the documentation [here](https://nemocas.github.io/AbstractAlgebra.jl/dev/ring_interface/)
> The results of deepcopy and all arithmetic operations, including powering and 
> division can be assumed to be new objects without other references being held,
> as can objects returned from constructors.

I don't think we should extend this to other random methods.

**Comment by tthsqe12** [comment]
I don't know. I was thinking of `real`, `imag` and `conj` as arithmetic operations. But, if they are not "on the list" then I will not copy here. Also, what counts as a constructor? It is well-known that all over the place the `R()` does not necessarily create a new object.

**Comment by tthsqe12** [comment]
I think is a good start. Pros:
1. diff coverage is 100%
2. there is no documentation added whatsoever
Cons:
1. It is currently slow. I will turn on the jets later.
2. Maybe it is worth documenting that the rings ZZi and QQi exist?

**Comment by thofma** [comment]
Yay! Let's get this in.

> 2. Maybe it is worth documenting that the rings ZZi and QQi exist?

Yes, please do this (in a separate PR).

**Pull request merged**